cpp/include/cudf/detail/timezone.cuh

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/timezone.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+
+namespace cudf::detail {
+
+/**
+ * @brief Returns the GMT offset for a given date and given timezone table.
+ *
+ * @param ttimes Transition times; trailing `cycle_entry_cnt` entries are used for all times
+ * beyond the one covered by the TZif file
+ * @param offsets Time offsets in specific intervals; trailing `cycle_entry_cnt` entries are used
+ * for all times beyond the one covered by the TZif file
+ * @param ts ORC timestamp
+ *
+ * @return GMT offset
+ */
+inline __device__ duration_s get_gmt_offset(table_device_view tz_table, timestamp_s ts)
+{
+  if (tz_table.num_rows() == 0) { return duration_s{0}; }
+
+  cudf::device_span<timestamp_s const> ttimes(tz_table.column(0).head<timestamp_s>(),
+                                              static_cast<size_t>(tz_table.num_rows()));
+
+  auto const ts_ttime_it = [&]() {
+    auto last_less_equal = [](auto begin, auto end, auto value) {
+      auto const first_larger = thrust::upper_bound(thrust::seq, begin, end, value);
+      // Return start of the range if all elements are larger than the value
+      if (first_larger == begin) return begin;
+      // Element before the first larger element is the last one less or equal
+      return first_larger - 1;
+    };
+
+    auto const file_entry_end = ttimes.begin() + (ttimes.size() - cycle_entry_cnt);
+
+    if (ts <= *(file_entry_end - 1)) {
+      // Search the file entries if the timestamp is in range
+      return last_less_equal(ttimes.begin(), file_entry_end, ts);
+    } else {
+      auto project_to_cycle = [](timestamp_s ts) {
+        static constexpr duration_s cycle_s =
+          cuda::std::chrono::duration_cast<duration_s>(duration_D{365 * cycle_years + (100 - 3)});
```

Comment:
ELI5: What's the calculation here? I'm afraid I don't follow.

Reply:
It codifies the leap year rules:


In our Gregorian calendar three criteria must be taken into account to identify leap years:
The year must be evenly divisible by 4;
If the year can also be evenly divided by 100, it is not a leap year;
unless...
The year is also evenly divisible by 400. Then it is a leap year.

I believe this translates to 100 (every fourth year) - 4 (every hundredth year ) + 1 (every four hundredth year) leap days in 400 years.

Reply:
Is this a rule we can get (or reference) from some other source? If we must implement it, is this codified in a way that is specific to ORC conventions or general to all time zone libraries?

To the greatest extent possible, cuDF should avoid implementing or defining its own rules for time math. We should always use (or reference) `libcudacxx`, other timezone libraries, an RFC, or similar.

Reply:
This is the definition of the Gregorian calendar. It's unrelated to ORC or timezones.
Given how specific this is (only multiples of 400 years have a fixed number of days, otherwise you need to know the year range) I doubt any library can compute this without computing number of days for each year individually (1-400 in this case, I think).

I can add a link to Gregorian calendar and clarification comments.

Reply:
I think I'm asking something different. I want a reference that states why we "search the 400-year cycle if outside of the file entries range" (in the comment below). We didn't come up with that, it's defined by some spec outside of libcudf.

Reply:
Ah, that. I actually don't know where this comes from. I'll try to find it and incorporate in some way.

Reply:
So, I've been digging around the interwebs and could not find anything on the use of 400 year cycle to compute time zone offsets in the future. 
The only thing I found is the name for this cycle - [Solar cycle](https://en.wikipedia.org/wiki/Solar_cycle_(calendar)). And even that is a seemingly overloaded term.
I added this to the names for the relevant constants, will incorporate it here as well.

Reply:
tried to make this clearer. Please let me know if it's still lacking